Add Compiler Visible Properties

vritant24 opened the pull request:
The Build properties `RootNamespace` and `ProjectDir` are needed for the new Namespace Analyzer in Roslyn to work outside of VS. Adding these properties to the Dotent SDK ensures that the analyzer works in dotnet build scenarios.

vritant24 committed:
Add Compiler Visible Properties

mavasani commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets:
```diff
@@ -102,4 +102,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   </ItemGroup>
+  
+  <!-- Build Properties Exposed to CodeStyle Analyzers -->
+  <ItemGroup Condition="$(EnforceCodeStyleInBuild)>
```
Are you missing quotes on the condition? Also, I am wondering if we should make this non-conditional so it enables this feature in IDE even when users are not enforcing code style on build. @jmarolf your thoughts?

vritant24 replied:
I have fixed the quotation. Would the addition in VS [here](https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/271377) not enable the IDE scenarios?

mavasani replied:
Hmm even with that PR, the condition disables it explicitly if user directly installs the CodeStyle NuGet package: https://github.com/dotnet/roslyn/blob/master/src/CodeStyle/CSharp/CodeFixes/build/Microsoft.CodeAnalysis.CSharp.CodeStyle.props

I think we should keep it simple:
1. Remove the condition here, so this feature works in VS + command line for SDK style projects for all the 3 analyzer sources:
   1. Default VS install with built-in code style analyzers coming from VS
   2. User sets `EnforceCodeStyleInBuild` to true and consumes code style analyzers from the .NET SDK
   3. User installs CodeStyle nuget package to consume code style analyzers from a specific version of the package.
2. Don't add support for this feature in non-SDK style projects, users would be asked to migrate to SDK style projects to get this feature.

mavasani replied:
With the above, you should not need the VS side PR. Also note that Microsoft.CodeAnalysis.Targets is not imported during `dotnet build`, so it will cause inconsistency where it works somewhere, but not everywhere.

vritant24 replied:
Ah I see. I wasn't aware this was explicitly disabled on installing the package. Removing the condition here makes sense.
On the note of Microsoft.CodeAnalysis.Targets, do we not intend to support Non-SDK style projects in VS since we wouldn't be able to support them in CLI build scenarios?

mavasani replied:
> do we not intend to support Non-SDK style projects in VS 

That is an open question that probably @mikadumont should answer.

vritant24 committed:
Fix quotation

vritant24 committed:
remove condition

mavasani commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets:
```diff
@@ -102,4 +102,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   </ItemGroup>
+  
+  <!-- Build Properties Exposed to CodeStyle Analyzers -->
+  <ItemGroup>
```
I think we want to condition the item group on this being C# or VB language.

jmarolf replied:
yep I think we want this: `<ItemGroup Condition="'$(Language)' == 'C#' Or '$(Language)' == 'VB'">`

vritant24 replied:
Added the condition to the ItemGroup

jmarolf commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets:
```diff
@@ -102,4 +102,10 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <!-- Build Properties Exposed to CodeStyle Analyzers -->
+  <ItemGroup>
+    <CompilerVisibleProperty Include="RootNamespace" />
+	  <CompilerVisibleProperty Include="ProjectDir" />
```
aliign with previous item

vritant24 committed:
Condition ItemGroup to Language

jmarolf commented:
@vritant24 which release do we want this to be in? sdk master is .NET 6. If you want this to ship in .NEt 5 you will need to rebase on top of `release/5.0.1xx-rc2`

vritant24 commented:
> 
> 
> @vritant24 which release do we want this to be in? sdk master is .NET 6. If you want this to ship in .NEt 5 you will need to rebase on top of `release/5.0.1xx-rc2`

Yes I think this is targeted to .NET 5, thanks. I've rebased and changed the target branch. correct me if I'm wrong @mavasani

genlu commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.Analyzers.targets:
```diff
@@ -103,4 +103,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   </ItemGroup>
+  
+  <!-- Build Properties Exposed to CodeStyle Analyzers -->
+  <ItemGroup Condition="'$(Language)' == 'C#' Or '$(Language)' == 'VB'">
```
If this will only used for sync namespace analyzer, then `'$(Language)' == 'VB'` would be unnecessary.

jmarolf replied:
why can't sync namespaces work for VB? both languages have namespaces?
